src/hotspot/share/opto/predicates.cpp

```diff
@@ -293,16 +301,43 @@ void InitializedAssertionPredicate::verify() const {
 }
 #endif // ASSERT
 
-// Initialized Assertion Predicates always have the dedicated OpaqueInitiailizedAssertionPredicate node to identify
-// them.
-bool InitializedAssertionPredicate::is_predicate(const Node* node) {
-  if (!may_be_assertion_predicate_if(node)) {
+// An Initialized Assertion Predicate always has a dedicated OpaqueInitializedAssertionPredicate node to identify it.
+bool InitializedAssertionPredicate::is_predicate(const Node* maybe_success_proj) {
+  if (!may_be_assertion_predicate_if(maybe_success_proj)) {
     return false;
   }
-  IfNode* if_node = node->in(0)->as_If();
-  return if_node->in(1)->is_OpaqueInitializedAssertionPredicate();
+  IfNode* if_node = maybe_success_proj->in(0)->as_If();
+  bool is_initialized_assertion_predicate = if_node->in(1)->is_OpaqueInitializedAssertionPredicate();
+  assert(!is_initialized_assertion_predicate || has_halt(maybe_success_proj->as_IfTrue()),
+         "Initialized Assertion Predicate must have a Halt Node on the failing path");
+  return is_initialized_assertion_predicate;
 }
 
+#ifdef ASSERT
+bool InitializedAssertionPredicate::has_halt(const IfTrueNode* success_proj) {
+  ProjNode* other_proj = success_proj->other_if_proj();
+  if (other_proj->outcnt() != 1) {
+    return false;
+  }
+
+  Node* out = other_proj->unique_out();
+  // Either the Halt node is directly the unique output.
+  if (out->is_Halt()) {
+    return true;
+  }
+  // Or we have a Region that merges serval paths to a single Halt node. Even though OpaqueInitializedAssertionPredicate
+  // nodes do not common up (i.e. NO_HASH), we could have Initialized Assertion Predicates from already folded loops
+  // being now part of the innermost loop. When then further splitting this loop, we could be cloning the If node
+  // of the Initialized Assertion Predicate (part of the loop body) while the OpaqueInitializedAssertionPredicate is not
+  // cloned because it's outside the loop body. We end up sharing the OpaqueInitializedAssertionPredicate between the
+  // original and the cloned If. This should be fine.
```

Comment:
Reason for not reusing `AssertionPredicate::has_halt()`. I will revisit the `AssertionPredicate` class in a future PR, so I have not tried to unify these methods in a way.